consensus/qkchash/qkchash.go

```diff
@@ -2,21 +2,95 @@ package qkchash
 
 import (
 	"encoding/binary"
+	"sort"
+	"sync"
+
 	"github.com/QuarkChain/goquarkchain/cluster/config"
 	"github.com/QuarkChain/goquarkchain/consensus"
 	"github.com/QuarkChain/goquarkchain/core/state"
 	"github.com/QuarkChain/goquarkchain/core/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	cacheEntryCnt   = 1024 * 64
+	cacheAheadRound = 64 // 64*30000
 )
 
+var (
+	EpochLength = uint64(30000) //blocks pre epoch
+)
+
+type cacheSeed struct {
+	mu     sync.RWMutex
+	caches []*qkcCache
+}
+
+func NewcacheSeed() *cacheSeed {
+	firstCache := generateCache(cacheEntryCnt, common.Hash{}.Bytes())
+	caches := make([]*qkcCache, 0)
+	caches = append(caches, firstCache)
+	return &cacheSeed{
+		caches: caches,
+	}
+}
+
+func (c *cacheSeed) getCacheFromHeight(block uint64) *qkcCache {
+	epoch := int(block / EpochLength)
+	lenCaches := len(c.caches)
+	if epoch < lenCaches {
+		return checkCache(c.caches[epoch])
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	needAddCnt := epoch - lenCaches + cacheAheadRound
+	seed := c.caches[len(c.caches)-1].seed
+	for i := 0; i < needAddCnt; i++ {
+		seed = crypto.Keccak256(seed)
+		c.caches = append(c.caches, generateCache(cacheEntryCnt, seed))
+	}
+	if needAddCnt > 2*cacheAheadRound {
+		for i := 0; i < needAddCnt-2*cacheAheadRound; i++ {
+			shrinkCache(c.caches[i])
+		}
+	}
+	return c.caches[epoch]
```

Comment:
What happen if the epoch's cache is shrunk (i.e., ls == nil), i.e., a client wants to get a cache from an old epoch?

Reply:
if the epoch is an old one, it will get from checkCache(c.caches[epoch]) which will fill in if the cache has been shrunk.
	if epoch < lenCaches {
		return checkCache(c.caches[epoch])
	}

Reply:
Then I believe there is a race condition between
1; after return c.caches[epoch] and before c.caches[epoch].ls is used; 2; the cache is evict, i.e., ls = nil.